Empower users to bring their own storage and file sources (#18127)
galaxyproject/galaxy: doc/source/admin/data.md

# Connecting Users and Data

Galaxy has countless ways for users to connect with things that might be considered their "data" - file sources (aka "remote files"), object stores (aka "storage locations"), data libraries, the upload API, visualizations, display applications, custom tools, etc...

> This document is going to discuss two of these (file sources and object stores) that are most important Galaxy administrators and how to build Galaxy configuration that allow administrators to let users tie into various pieces of infrastructure.

Reviewer: ```suggestion
This document is going to discuss two of these (file sources and object stores) that are most important for Galaxy administrators and how to build Galaxy configuration that allows administrators to let users tie into various pieces of infrastructure.
```

```{contents} Table of Contents
:depth: 4
```

## Datasets vs Files

File sources in Galaxy are a sprawling concept but essentially they provide users access to simple files (stored hierarchically into folders) that can be navigated and imported into Galaxy. Importing a "file" into Galaxy generally creates a copy of that file into a Galaxy "object store". Once these files are stored in Galaxy,
they become "datasets". A Galaxy dataset is much more than a simple file - Galaxy datasets include various generic metadata a datatype, datatype specific metadata, and ownership and sharing rules managed by Galaxy.

> Galaxy object stores (called "storage locations" in the UI) store datasets and global (accessible to all users) object stores are configured with the ``galaxy.yml`` property ``object_store_config_file`` (or  ``object_store_config`` for a configuration embedded right in ``galaxy.yml``) that defaults to ``object_store_conf.xml`` or ``object_store_conf.yml`` if either is present in Galaxy's configuration directory. Galaxy file sources provide users access to raw files and global files sources are configured with the ``galaxy.yml`` property ``file_sources_config_file`` (or ``file_sources`` for embedded configurations) that defaults to ``file_sources_conf.yml`` if that file is present in Galaxy's configuration directory.

Reviewer: For the sake of simplicity should we just not document the XML config file?

Author: My PR itself has XML examples 😆 . I don't think we're there yet.

Some of Galaxy's most updated and complete administrator documentation can be found in configuration sample files - this is definitely the case for object stores and file sources. The relevant sample configuration files include  [file_sources_conf.yml.sample](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/sample/file_sources_conf.yml.sample) and [object_store_conf.sample.yml](https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/sample/object_store_conf.sample.yml).

> File sources and object stores configured with the above files essentially are available to all users of your Galaxy instance - hence this document describes them as "global" file sources and object  stores. File source configurations do allow some templating that does allow the a global file source to be materialized differently for different users. For instance, you as an admin may setup a Dropbox file source and may explicitly add custom user properties that allow that single Dropbox file source to read from a user's preferences. Since there is just one Dropbox service and most people only have a single Dropbox account, this use case can be somewhat adequately addressed by the global file source and the global user preferences file. For a use case like Amazon S3 buckets though for instance, a single bucket file source that is parameterized one way is probably more clearly inadequate. For instance, users would very likely want to attach different buckets for different projects.  Additionally, the Galaxy user interface doesn't tie the user preferences to the particular file source and so this method introduces a huge education burden on your Galaxy instance. Finally, the templating available to file sources are not available for object stores - and allowing users to describe how they would like datasets stored and to pay for their own dataset storage are important use cases.

Reviewer: ```suggestion
File sources and object stores configured with the above files essentially are available to all users of your Galaxy instance - hence this document describes them as "global" file sources and object stores. File source configurations do allow some templating that does allow the global file source to be materialized differently for different users. For instance, you as an admin may setup a Dropbox file source and may explicitly add custom user properties that allow that single Dropbox file source to read from a user's preferences. Since there is just one Dropbox service and most people only have a single Dropbox account, this use case can be somewhat adequately addressed by the global file source and the global user preferences file. For a use case like Amazon S3 buckets though for instance, a single bucket file source that is parameterized one way is probably more clearly inadequate. For instance, users would very likely want to attach different buckets for different projects.  

Additionally, the Galaxy user interface doesn't tie the user preferences to the particular file source and so this method introduces a huge education burden on your Galaxy instance. Finally, the templating available to file sources is not available for object stores - and allowing users to describe how they would like datasets stored and to pay for their own dataset storage are important use cases.
```

I'm wondering if there is really a valid use-case where the global file-source templating - bound to user-preferences - is now still valid. Imho we should discourage this from now on and tell people to use the user-based-file sources, also for dropbox. Its much cleaner for the user and easier for admins.

Reviewer: Alternatively, we could leave this config option out of this text and just mention it below this page as an alternative for special situations.

Reviewer: If you agree, I'm happy to try to disentangle it a bit.

Author: I will admit this language was added... defensively(?)... to validate the existence of this work. I was worried there would be people who wouldn't understand the point of it. It is really validating that you want to remove that language.

The vault documentation has examples of the global templating and I haven't looked but I would imagine the training material does as well? I would say we get this stuff merged and we get it working in production for a year and then circle back and decide if we want to downplay the global templating version of all of this. This should be in a couple releases before it is considered the best practice for some of those use cases I would say?

can apply to both file source and object store plugins, and the user interface is unified from the
template configuration file (you as the admin do not need to explicitly declare user preferences and
your users do not need to navigate seemingly unrelated preferences to get plugins to work).

## Object Store Templates

Galaxy's object store templates are configured as a YAML list of template objects. This list
can be placed ``object_store_templates.yml`` in Galaxy configuration directory (or any path
pointed to by the configuration option ``object_store_templates_config_file`` in ``galaxy.yml``).
Alternatively, the  configuration can be placed directly into ``galaxy.yml`` using the
``object_store_templates`` configuration option.

A minimal object store template might look something like:

```{literalinclude} ../../../lib/galaxy/objectstore/templates/examples/simple_example.yml
:language: yaml
```

### Object Store Types

#### ``disk``

This is the most basic sort of object store template that just makes disk paths available to users
for storing data. Paths can be built up from the user supplied variables, user details, supplied

> environment variables, etc.. The simple example used to demonstrate these concepts just uses

Reviewer: ```suggestion
environment variables, etc.. The simple example just uses
```

> a user supplied project name and the user's username to produce a unique path for each user

Reviewer: Q: what happens when a user changes its username? The username is unique I assume, but can be changed.

:language: yaml
```

#### Allow Users to Define AWS S3 Buckets as Object Stores

```{literalinclude} ../../../lib/galaxy/objectstore/templates/examples/production_aws_s3.yml
:language: yaml
```

![Screenshot](user_object_store_form_full_aws_s3.png)

#### Allow Users to Define Google Cloud Provider S3 Interop Storage Buckets as Object Stores

This template includes descriptions of how to generate HMAC keys used by this interoperability
layer provided by Google and lots of links to relevant Google Cloud Storage documentation.

```{literalinclude} ../../../lib/galaxy/objectstore/templates/examples/production_gcp_s3.yml
:language: yaml
```

![Screenshot](user_object_store_form_full_gcp_s3_interop.png)

## File Source Templates

Galaxy's file source templates are configured as a YAML list of template objects. This list
can be placed ``file_source_templates.yml`` in Galaxy configuration directory (or any path
pointed to by the configuration option ``file_source_templates_config_file`` in ``galaxy.yml``).
Alternatively, the  configuration can be placed directly into ``galaxy.yml`` using the
``file_source_templates`` configuration option.

### File Source Types

> #### ``posix``

Reviewer: Is there a reason for calling this `posix` in contrast to `disk`, besides historical reasons?

```{literalinclude} ../../../lib/galaxy/files/templates/examples/production_ftp.yml
:language: yaml
```

The syntax for the ``configuration`` section of ``ftp`` templates looks like this.

![](file_source_ftp_configuration_template.png)

At runtime, after the ``configuration`` template is expanded, the resulting dictionary
passed to Galaxy's file source plugin infrastructure looks like this and should match a subset
of what you'd be able to add directly to ``file_sources_conf.yml`` (Galaxy's global file source
configuration).

![](file_source_ftp_configuration.png)

#### ``azure``

The syntax for the ``configuration`` section of ``azure`` templates looks like this.

![](file_source_azure_configuration_template.png)

At runtime, after the ``configuration`` template is expanded, the resulting dictionary
passed to Galaxy's file source plugin infrastructure looks like this and should match a subset
of what you'd be able to add directly to ``file_sources_conf.yml`` (Galaxy's global file source
configuration).

![](file_source_azure_configuration.png)

### YAML Syntax

![galaxy.files.templates.models](file_source_templates.png)

### Ready To Use Production File Source Templates

> The templates have been tested by a Galaxy developer and are sufficiently generic that

Reviewer: ```suggestion
The templates are sufficiently generic that
```

Author: Yeah? 😆 . I'll make the change but it seems odd.

they may make sense for a variety of Galaxy instances, address a variety of potential use

> cases, and do not have need any additional tailoring, parameterization, or other

Reviewer: ```suggestion
cases, and do not need any additional tailoring, parameterization, or other
```

> customization. These assume your Galaxy instance has a Vault configured and you're

Reviewer: link to the Galaxy Vault docs?

Many large instances of Galaxy are configured with Ansible and much of the existing administrator
documentation leverages Ansible. The configuration template files using Jinja templating and so
does Ansible by default. This might result in a lack of clarity of when templates (strings 
starting with ``{{`` and ending with ``}}``) are being evaluated. Ansible templates are evaluated
at deploy time and the configuration objects describing plugins are evaluated at Galaxy runtime.

The easiest way to fix this is probably to store these templates files in your Ansible as plain files
and not templates. If you'd like to use Ansible templating to build up these files you'll very
likely need to tell either Galaxy or Ansible to use something other than ``{{`` and ``}}`` for 
templating variables. This can be done by placing a directive at the top of your template that
is consumed by Ansible. For instance, to have ``[%`` and ``%]`` used instead of ``{{`` and ``}}``
by Ansible at deploy time, the file could start with:

```
#jinja2:variable_start_string:'[%' , variable_end_string:'%]'
```

In this case, variables wrapped by ``[%`` and ``%]`` are expanded by Ansible and use the Ansible

> environment and ``{`` and ``}`` are reserved for Galaxy templating.

Reviewer: ```suggestion
environment and ``{{`` and ``}}`` are reserved for Galaxy templating.
```

?

```{literalinclude} ../../../lib/galaxy/files/templates/examples/admin_secrets_with_defaults.yml
:language: yaml
```

### ``user``

This dictionary object exposes information about user configuring and using a target template
configuration. These values are populated from the ``galaxy_user`` table of the Galaxy database.
The current properties exposed include:

| Key          | Description                                               | 
|--------------|-----------------------------------------------------------|
| ``username`` | string corresponding the username of the Galaxy user      |
| ``email``    | string corresponding the email of the Galaxy user         |
| ``id``       | integer primary key of user object in the Galaxy database |

The simple example of project scratch storage used to describe these concepts made
use the Galaxy user's username to generate unique paths.

```{literalinclude} ../../../lib/galaxy/objectstore/templates/examples/simple_example.yml
:language: yaml
```

### ``ensure_path_component``

This [Jinja filter](https://jinja.palletsprojects.com/en/3.0.x/templates/#filters)

> will fail template evaluation if the value it is applies to is not

Reviewer: ```suggestion
will fail template evaluation if the value it is applied to is not
```
